File: advance/src/eu/advance/logistics/flow/engine/api/ds/AdvanceWebDataSource.java
/*
 * Copyright 2010-2013 The Advance EU 7th Framework project consortium
 *
 * This file is part of Advance.
 *
 * Advance is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as
 * published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * Advance is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with Advance.  If not, see
 * <http://www.gnu.org/licenses/>.
 *
 */

package eu.advance.logistics.flow.engine.api.ds;

import hu.akarnokd.reactive4java.base.Func0;

import java.net.MalformedURLException;
import java.net.URL;

import org.slf4j.LoggerFactory;

import eu.advance.logistics.flow.engine.api.core.Copyable;
import eu.advance.logistics.flow.engine.api.core.HasPassword;
import eu.advance.logistics.flow.engine.api.core.Identifiable;
import eu.advance.logistics.flow.engine.xml.typesystem.XElement;
import eu.advance.logistics.flow.engine.xml.typesystem.XSerializable;

/**
 * The web data source configuration record.
 * @author akarnokd, 2011.09.20.
 */
public class AdvanceWebDataSource extends AdvanceCreateModifyInfo 
implements XSerializable, HasPassword, Copyable<AdvanceWebDataSource>, Identifiable<String> {
	/** The data source name as used by the blocks. */
	public String name;
	/** The URL. */
	public URL url;
	/** The login type enumeration. */
	public AdvanceLoginType loginType;
	/** The keystore name if the loginType is CERTIFICATE. */
	public String keyStore;
	/** The user or key alias name. */
	public String userOrKeyAlias;
	/**
	 * The password for BASIC login or the key password for the CERTIFICATE.
	 * <p>Note that passwords are never returned from the 
	 * control API calls and are always {@code null}.</p> 
	 * <p>An empty password should be an empty {@code char} array. To keep
	 * the current password, use {@code null}.</p>
	 * @return  
	 */
	private char[] password;
	/** The function to create a new instance of this class. */
	public static final Func0<AdvanceWebDataSource> CREATOR = new Func0<AdvanceWebDataSource>() {
		@Override
		public AdvanceWebDataSource invoke() {
			return new AdvanceWebDataSource();
		}
	};
	@Override
	public void load(XElement source) {
		name = source.get("name");
		try {
			url = new URL(source.get("url"));
		} catch (MalformedURLException ex) {
			LoggerFactory.getLogger(AdvanceWebDataSource.class).error(ex.toString(), ex);
		}
		loginType = AdvanceLoginType.valueOf(source.get("login-type"));
		keyStore = source.get("keystore");
		userOrKeyAlias = source.get("user-or-key");
		password = getPassword(source, "password");
		
		super.load(source);
	}
	@Override
	public void save(XElement destination) {
		
		destination.set("name", name);
		destination.set("url", url);
		destination.set("login-type", loginType);
		destination.set("keystore", keyStore);
		destination.set("user-or-key", userOrKeyAlias);
		setPassword(destination, "password", password);
		
		
		super.save(destination);
	}
	@Override
	public AdvanceWebDataSource copy() {
		AdvanceWebDataSource result = new AdvanceWebDataSource();
		
		result.name = name;
		result.url = url;
		result.loginType = loginType;
		result.keyStore = keyStore;
		result.userOrKeyAlias = userOrKeyAlias;
		result.password = password != null ? password.clone() : null;
		
		assignTo(result);
		
		return result;
	}
	@Override
	public char[] password() {
		return password != null ? password.clone() : null;
	}
	@Override
	public void password(char[] newPassword) {
		password = newPassword != null ? newPassword.clone() : null;
	}
	@Override
	public String id() {
		return name;
	}
}

File: advance-api-cs/AdvanceAPIClient/Classes/AdvanceWebDataSource.cs
/*
 * Copyright 2010-2013 The Advance EU 7th Framework project consortium
 *
 * This file is part of Advance.
 *
 * Advance is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as
 * published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * Advance is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with Advance.  If not, see
 * <http://www.gnu.org/licenses/>.
 *
 */
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Globalization;
using System.Xml;

using AdvanceAPIClient.Core;
using AdvanceAPIClient.Interfaces;

namespace AdvanceAPIClient.Classes
{
    public class AdvanceWebDataSource : AdvanceCreateModifyInfo, IPassword, ICopiable<AdvanceWebDataSource>, IIdentifiable<string>
    {
        /// <summary>
        /// name used by blocks to reference this data source
        /// </summary>
        public string Id { get { return this.Name; } }
        public string Name;

         /// <summary>
        /// User's password.
        /// <remarks>
        /// Note that passwords are never returned from the 
        /// control API calls and are always null
        /// An empty password should be an char[0] array. To keep
        /// the current password, use null
        /// </remarks>
        /// </summary>
        public char[] Password
        {
            get { return (this.password == null) ? null : (char[])this.password.Clone(); }
            set { this.password = (value == null) ? null : (char[])value.Clone(); }
        }
        private char[] password;

        /// <summary>
        ///  url
        /// </summary>
        public Uri Url;
        /// <summary>
        /// login type
        /// </summary>
        public AdvanceLoginType LoginType;
        /// <summary>
        /// Keystore name if the loginType is CERTIFICATE.
        /// </summary>
        public string KeyStore;
        /// <summary>
        /// User or key alias name
        /// </summary>
        public string UserOrKeyAlias;

        public AdvanceWebDataSource() : base() { }

        /// <summary>
        /// Fill object data from xml node
        /// </summary>
        /// <param name="source">Source Xml</param>
        protected override void LoadFromXmlNode(XmlNode source)
        {
            base.LoadFromXmlNode(source);
            this.Name = GetAttribute(source, "name");
            this.Url = GetUriAttribute(source, "url");
            this.UserOrKeyAlias = GetAttribute(source, "user-or-key", null);
            this.password = GetPassword(source, "password");
            this.KeyStore = GetAttribute(source, "keystore", null);
            this.LoginType = GetEnumAttribute<AdvanceLoginType>(source, "login-type", AdvanceLoginType.NONE);
	    }

        /// <summary>
        /// Fills  object's xml
        /// </summary>
        /// <param name="node">Empty node for the object with name "tagname"</param>
        protected override void FillXmlElement(XmlElement node)
        {
            base.FillXmlElement(node);
            AddAttribute(node, "name", this.Name);
            AddAttribute(node, "url", this.Url);
            AddAttribute(node, "user-or-key", this.UserOrKeyAlias);
            AddAttribute(node, "keystore", this.KeyStore);
            AddAttribute(node, "login-type", this.LoginType);
            SetPassword(node, "password", password);
        }

        public AdvanceWebDataSource Copy()
        {
            AdvanceWebDataSource result = new AdvanceWebDataSource();
            this.AssignTo(result);
            result.Name = this.Name;
            result.Url = this.Url;
            result.UserOrKeyAlias = this.UserOrKeyAlias;
            result.password = this.Password;
            result.KeyStore = this.KeyStore;
            result.LoginType = this.LoginType;
            return result;
        }

    }
}

